src/x86_64/schedule.c

```diff
@@ -158,8 +158,12 @@ NOTRACE void __attribute__((noreturn)) runloop_internal()
         kern_unlock();
     }
 
+    if (ci->current_thread)
```

Comment:
Yes, indeed that's sufficient; I removed the call to thread_resume() from runloop_internal().

Reply:
I still don't see why there is a call to thread_pause() before kernel_sleep() in runloop_internal. If we get to that point, no thread is being scheduled on the CPU anyway, and we're about to sti; hlt. Shouldn't the thread have already been paused (on a syscall or something else that made it not runnable)?

Reply:
If we remove that call to thread_pause(), there is nothing that calls thread_pause() when a thread is going to sleep (unless there is another thread ready to run so that the CPU switches directly from one thread to the other).

Reply:
Yes, there needs to be a call to thread_pause when system processing on behalf of the thread has ceased, though I think it should happen before the call to runloop, or at least at the top of the runloop (but again, it gets murky when specific callbacks get pulled into runloop - we've had this problem in the past before we reorganized scheduling). Otherwise, time spent serving items the bhqueue and runqueue (including things like lwIP processing and other non-thread-specific tasks) is accounted on behalf of the thread.

It seems that we need a little gate to mark the end of system processing on behalf of a thread - maybe a wrapper around the call to runloop in a few places (thread_sleep_*, signal delivery, syscall_schedule, etc.).

Actually, in lieu of a wrapper...I'm starting to think maybe adding a 'pause' method to the frame might be a cleaner and more general way to go about this. If set, it would just get called unconditionally at the top of runloop (and technically at the top of common_handler, too, though only if it is known that the handler will terminate with a call to runloop and not a frame_return (e.g. page fault)).

I can take a crack and PR if it pans out.